crates/router/src/core/payments/helpers.rs

```diff
@@ -2060,27 +2060,50 @@ pub fn generate_mandate(
 }
 
 // A function to manually authenticate the client secret with intent fulfillment time
-pub fn authenticate_client_secret(
+pub async fn authenticate_client_secret(
     request_client_secret: Option<&String>,
     payment_intent: &PaymentIntent,
     merchant_intent_fulfillment_time: Option<i64>,
+    db: &dyn StorageInterface,
 ) -> Result<(), errors::ApiErrorResponse> {
     match (request_client_secret, &payment_intent.client_secret) {
         (Some(req_cs), Some(pi_cs)) => {
             if req_cs != pi_cs {
                 Err(errors::ApiErrorResponse::ClientSecretInvalid)
             } else {
                 //This is done to check whether the merchant_account's intent fulfillment time has expired or not
-                let payment_intent_fulfillment_deadline =
-                    payment_intent.created_at.saturating_add(Duration::seconds(
-                        merchant_intent_fulfillment_time
-                            .unwrap_or(consts::DEFAULT_FULFILLMENT_TIME),
-                    ));
-                let current_timestamp = common_utils::date_time::now();
-                fp_utils::when(
-                    current_timestamp > payment_intent_fulfillment_deadline,
-                    || Err(errors::ApiErrorResponse::ClientSecretExpired),
-                )
+                let payment_link_db = if let Some(payment_link_id) = &payment_intent.payment_link_id
+                {
+                    db.find_payment_link_by_payment_link_id(
+                        payment_link_id,
+                        &payment_intent.merchant_id,
+                    )
+                    .await
+                    .ok()
+                } else {
+                    None
+                };
+                if let Some(payment_link_db) = payment_link_db {
+                    let current_timestamp = common_utils::date_time::now();
+                    fp_utils::when(
+                        current_timestamp
+                            >= payment_link_db
+                                .fullfilment_time
+                                .unwrap_or(current_timestamp),
+                        || Err(errors::ApiErrorResponse::ClientSecretExpired),
+                    )
+                } else {
+                    let payment_intent_fulfillment_deadline =
+                        payment_intent.created_at.saturating_add(Duration::seconds(
+                            merchant_intent_fulfillment_time
+                                .unwrap_or(consts::DEFAULT_FULFILLMENT_TIME),
```

Comment:
The logic is dependent on the fulfillment time, which can either come from `payment_intent` or `payment_link` (which overrides the payment_intent fulfillment time), can you refactor this based on this idea?